frontend-react/src/shared/Tooltip.tsx

```diff
@@ -0,0 +1,215 @@
+}
+
+export function useTooltip({
+    initialOpen = false,
+    placement = "top",
+    open: controlledOpen,
+    onOpenChange,
+    offsetBy,
+}: TooltipOptions = {}) {
+    const [uncontrolledOpen, setUncontrolledOpen] = React.useState(initialOpen);
+    const isControlled = controlledOpen != null;
+    const open = controlledOpen ?? uncontrolledOpen;
+    const setOpen = React.useCallback(
+        (isOpen: boolean) => {
+            console.log("ding");
+            onOpenChange?.(isOpen);
+            if (controlledOpen == null) {
+                setUncontrolledOpen(isOpen);
+            }
+        },
+        [onOpenChange, controlledOpen]
+    );
+
+    const data = useFloating({
+        placement,
+        open,
+        onOpenChange: setOpen,
+        whileElementsMounted: autoUpdate,
+        middleware: [
+            flip({
+                fallbackAxisSideDirection: "start",
+            }),
+            shift({ padding: 5 }),
+            offset(offsetBy),
+        ],
+    });
+
+    const context = data.context;
+
+    const hover = useHover(context, {
+        move: false,
+        enabled: !isControlled,
+    });
+    const focus = useFocus(context, {
+        enabled: !isControlled,
+    });
+    const dismiss = useDismiss(context);
+    const role = useRole(context, { role: "tooltip" });
+
+    const interactions = useInteractions([hover, focus, dismiss, role]);
+
+    return React.useMemo(
+        () => ({
+            open,
+            setOpen,
+            ...interactions,
+            ...data,
+        }),
+        [open, setOpen, interactions, data]
+    );
+}
+
+export type TooltipContextType = ReturnType<typeof useTooltip> | null;
+
+export const TooltipContext = React.createContext<TooltipContextType>(null);
+
+export const useTooltipContext = () => {
+    const context = React.useContext(TooltipContext);
+
+    if (context == null) {
+        throw new Error("Tooltip components must be wrapped in <Tooltip />");
+    }
+
+    return context;
+};
+
+export type TooltipProps = { children: React.ReactNode } & TooltipOptions;
+
+/**
+ * Prepares the tooltip context for the provided children components. Allows for either controlled or
+ * uncontrolled tooltips via the onOpenChange prop. You must provide TooltipTrigger and TooltipContent
+ * wrapped components and only as children within a tooltip context (which this creates).
+ * ```
+ * <Tooltip>
+ *   <TooltipTrigger>Hover me!</TooltipTrigger>
+ *   <TooltipContent>Hello!</TooltipContent>
+ * </Tooltip>
+ * ```
+ */
+export function Tooltip({ children, ...options }: TooltipProps) {
+    // This can accept any props as options, e.g. `placement`,
+    // or other positioning options.
+    const tooltip = useTooltip(options);
+    return (
+        <TooltipContext.Provider value={tooltip}>
+            {children}
+        </TooltipContext.Provider>
+    );
+}
+
+export type TooltipTriggerPropsBase = React.HTMLProps<HTMLElement> & {
+    asChild?: boolean;
```

Comment:
Could we just pass in the forward ref as a prop instead of specifying a boolean and then children?  Not sure if that'd be better or worse, honestly -- just a thought!  😬